jax/_src/nn/functions.py

```diff
@@ -765,3 +768,203 @@ def hard_silu(x: ArrayLike) -> Array:
+
+  Computes the attention function on Quary, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`. This function supports
+  optional or custom (1) scale function over :code:`logits`; (2) attention mask
+  over scaled :code:`logits`; (3) softmax over masked :code:`logits`; (4)
+  dropout over :code:`probs`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+          :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. We use additive
+          masking: 0 represents true and values below a large negative number (
+          e.g. :code:`-0.5 * jnp.finfo(dtype).max`); shape broadcastable to
+          :code:`(BNTS)`.
+    scale: scale for the logits. If None, the scale will be set to 1 divided by
+           the square root of query size.
+    logits_cap: caps the logits if it is > 0.0. Defaults to None, meaning no
+                capping. Otherwise, it applies
+                :code:`cap * jnp.tanh(logits / cap)`.
+    mask_fn: custom masking function takes in :code:`logits` and :code:`mask`
+             and outputs masked :code:`logits`.
+    softmax_fn: custom softmax function takes in :code:`logits` and outputs
+                :code:`probs`.
+    dropout_fn: custom dropout function takes in :code:`probs` outputs the new
+                `probs`.
+    _qk_dot_general: custom dot function for qk dot (mainly for quantization).
+    _pv_dot_general: custom dot function for pv dot (mainly for quantization).
+
+  Returns:
+    An array with the same shape of :code:`query`. If flash attention is not
+    used, the `probs` array will also be returned.
+  """
+  def _assert_has_shape(t: ArrayLike, shape: Sequence[int]) -> None:
+    assert t.ndim == len(shape)
+    value_str1 = f't.shape={t.shape}'
+    value_str2 = f'shape={shape}'
+    for i in range(t.ndim):
+      if shape[i] != -1:
+        assert (
+            t.shape[i] == shape[i]
+        ), f'Shapes are not equal {value_str1} vs {value_str2}'
+    
+  B, S, N, H = key.shape
+  _assert_has_shape(value, [B, S, N, H])
+  _assert_has_shape(query, [B, -1, N, H])
+  T = query.shape[1]
+  scale_val = (1.0 / np.sqrt(H) if scale is None else scale)
+
+  if not (mask is None or mask_fn is None):
+      raise ValueError("Cannot provide both mask and mask_fn")
+
+  # Try if the flash attention can be used. If failed, fall back to the default 
```

Comment:
Users of this function will very likely want to use an optimized FlashAttention kernel. Yet its extremely opaque to the user exactly what they must do to actually use a fast kernel (what shapes? what API surface of this function are they allowed to use?)

Reply:
Currently, you are not throwing an error when flashattention is not used. I think its easy to miss error messages that are printed/logged somewhere. This is related to the idea of having an implementation option, where users can explicitly ask for the cudnn implementation and fail if its not supported (with good errors).